File: src/main/java/org/example/BatchTester.java
package org.example;

import java.util.Optional;

import static org.example.ValidateTester.validateFor;

public class BatchTester extends Tester {
    @Override
    protected boolean validate( Product product ) {
        return product.batch() >= 1000 && product.batch() <= 2000
                ? Optional.ofNullable( getNextTester() )
                          .map( validateFor( product ) )
                          .orElse( true )
                : false;
    }

}

File: src/main/kotlin/BatchTester.kt
import extensions.isInBetween

class BatchTester : Tester() {
    override fun test(product: Product): Boolean =
        product.validate()


    private fun Product.validate() =
        if (batch.isInBetween(1000, 2000))
            nextTester.test(this)
        else false
}